WordPress/gutenberg
# Refactor dropzone for unmodified default blocks with `useBlockDropZone` and `InsertionPoint`

## What?

A follow-up and refactor of #42722, an alternative to #44606.

This refactor basically reverts the implementation in the Paragraph block (#42722) and re-implements the same feature in `useBlockDropZone` and `<InsertionPoint>` using 
`<BlockPopover>` and `isUnmodifiedDefaultBlock` (as suggested in https://github.com/WordPress/gutenberg/pull/42722#issuecomment-1262025220). This refactor allows us to reuse the same drop zone and their event listeners so that we don't have to bind them elsewhere.

## Why?


The original PR fails some performance metrics when rendering a large list of empty paragraph blocks and it doesn't take `setDefaultBlock` into account. This PR solves both issues.

## How?


1. We calculate the drop indicator and take unmodified default blocks into account in `useBlockDropZone`.
2. If there's an unmodified default block nearby then change the drop `operation` to `replace` and call `showInsertionPoint` with it.
3. Render the dropzone indicator (fully presentational component) on top of the unmodified default block using `<BlockPopover>`.

This PR has a nice side effect that dragging in-between unmodified default blocks will also trigger the drop zone indicator and allow dropping. This is smoother than the previous implementation, where only hovering on the empty paragraph blocks will trigger the drop zone (which is often very small). See it in action in the below screencast.

## Testing Instructions




1. Create some empty paragraph blocks (unmodified default blocks) and some other blocks with contents.
2. Drag something around the post and see that the dropzone indicator appears when hovering on the empty paragraph blocks.
3. Drop to replace the empty paragraph blocks with some blocks (depending on the dragged data).

## Screenshots or screencast 

https://user-images.githubusercontent.com/7753001/193731962-73874eca-a7ef-42bb-b06c-714388b507c8.mp4

## How to review

1. See `useBlockDropZone` first, and follow the logic there.
2. `block-editor`'s actions (`showInsertionPoint`) -> reducer.
3. `<InsertionPoint>` is where the presentational component renders.
4. `<BlockPopoverDropZone>` (`block-popover/drop-zone.js`) is where the presentational drop zone component is.
5. Other source files are mostly code removals from #42722.
6. The rests are tests and renamings.
